mmaction/utils/module_hooks.py

```diff
@@ -0,0 +1,75 @@
+import torch
+from mmcv.utils import Registry, build_from_cfg
+
+MODULE_HOOKS = Registry('module_hooks')
+
+
+def register_module_hooks(Module, module_hooks_list):
+    handles = []
+    for module_hook_cfg in module_hooks_list:
+        hook_pos = module_hook_cfg.pop('hook_pos', 'forward_pre')
+        if hook_pos == 'forward_pre':
+            handle = Module.register_forward_pre_hook(
+                build_from_cfg(module_hook_cfg, MODULE_HOOKS).hook_func())
+        elif hook_pos == 'forward':
+            handle = Module.register_forward_hook(
+                build_from_cfg(module_hook_cfg, MODULE_HOOKS).hook_func())
+        elif hook_pos == 'backward':
+            handle = Module.register_backward_hook(
+                build_from_cfg(module_hook_cfg, MODULE_HOOKS).hook_func())
+        else:
+            raise ValueError(
+                f'hook_pos must be `forward_pre`, `forward` or `backward`, '
+                f'but get {hook_pos}')
+        handles.append(handle)
+    return handles
+
+
+@MODULE_HOOKS.register_module()
+class GPUNormalize:
+    """Normalize images with the given mean and std value on GPUs.
+
+    Call the member function ``hook_func`` will return the forward pre-hook
+    function for module registration.
+
+    Args:
+        mean (Sequence[float]): Mean values of different channels.
+        std (Sequence[float]): Std values of different channels.
+    """
+
+    def __init__(self, input_format, mean, std):
+        if input_format not in ['NCTHW', 'NCHW', 'NCHW_Flow', 'NPTCHW']:
+            raise ValueError(f'The input format {input_format} is invalid.')
+        self.input_format = input_format
+        _mean = torch.tensor(mean)
+        _std = torch.tensor(std)
+        if input_format == 'NCTHW':
+            self._mean = _mean[None, :, None, None, None]
+            self._std = _std[None, :, None, None, None]
+        elif input_format == 'NCHW':
+            self._mean = _mean[None, :, None, None]
+            self._std = _std[None, :, None, None]
+        elif input_format == 'NCHW_Flow':
+            self._mean = _mean[None, :, None, None]
+            self._std = _std[None, :, None, None]
+        elif input_format == 'NPTCHW':
+            self._mean = _mean[None, None, None, :, None, None]
+            self._std = _std[None, None, None, :, None, None]
+
+    def hook_func(self):
+
+        def normalize_hook(Module, input):
+            x = input[0]
+            if not hasattr(self, 'mean'):
```

Comment:
Have you tested on the normal, **non**-distribued DataParallel?

Reply:
I guess this is doesn't work with non-distributed DataParallel since the module data only stored in cuda:0.

Actually, MMDataParallel has not supported training on multi-GPUs. Do we still need to test it in MMDataParallel?
![image](https://user-images.githubusercontent.com/35596075/107481204-c9e03280-6bb8-11eb-8d92-6dce9ea28ec7.png)

Reply:
If we do not cache the tensor in self.mean and self.std, and call the `.to(device)` every time, will the speed be affected?

Reply:
There are no free GPUs at the moment. I'll see the result later

Reply:
The speed is almost unaffected